analysis/metric_space.lean

```diff
@@ -585,6 +615,169 @@ instance prod.metric_space_max [metric_space β] : metric_space (α × β) :=
   end,
   to_uniform_space := prod.uniform_space }
 
+lemma prod.dist_eq [metric_space β] {x y : α × β} :
+  dist x y = max (dist x.1 y.1) (dist x.2 y.2) := rfl
+
+end prod
+
+section sum
+variables [metric_space β] [inhabited α] [inhabited β]
+
+/--Distance on a disjoint union. There are many (noncanonical) ways to put a distance compatible with each factor.
+If the two spaces are bounded, one can say for instance that each point in the first is at distance
+`diam α + diam β + 1` of each point in the second.
+Instead, we choose a construction that works for unbounded spaces, but requires basepoints.
+We embed isometrically each factor, set the basepoints at distance 1,
+arbitrarily, and say that the distance from `a` to `b` is the sum of the distances of `a` and `b` to
+their respective basepoints, plus the distance 1 between the basepoints. -/
+private def sum.dist : α ⊕ β → α ⊕ β → ℝ
+| (sum.inl a) (sum.inl a') := dist a a'
+| (sum.inr b) (sum.inr b') := dist b b'
+| (sum.inl a) (sum.inr b)  := dist a (default α) + 1 + dist (default β) b
+| (sum.inr b) (sum.inl a)  := dist b (default β) + 1 + dist (default α) a
+
+instance : has_dist (α ⊕ β) :=
+{ dist := sum.dist }
+
+lemma sum.dist_eq {x y : α ⊕ β} :
+  dist x y = sum.dist x y := rfl
+
+lemma sum.one_dist_le {x : α} {y : β} :
+  1 ≤ dist (@sum.inl α β x) (@sum.inr α β y) :=
+begin
+  have : 0 + 0 ≤ dist x (default α) + dist (default β) y := add_le_add dist_nonneg dist_nonneg,
+  simpa [dist, sum.dist] using this
+end
+
+lemma sum.one_dist_le' {x : α} {y : β} :
+  1 ≤ dist (@sum.inr α β y) (@sum.inl α β x) :=
+begin
+  have : 0 + 0 ≤ dist y (default β) + dist (default α) x := add_le_add dist_nonneg dist_nonneg,
+  simpa [dist, sum.dist] using this
+end
+
+open sum
+
+/--The distance on the disjoint union indeed defines a metric space. All the distance properties follow from our
+choice of the distance. The harder work is to show that the uniform structure defined by the distance coincides
+with the disjoint union uniform structure.-/
+instance : metric_space (α ⊕ β) :=
```

Comment:
I don't like this being an instance, because of the arbitrary choices involved. A more canonical instance is the emetric space that puts the factors infinitely far apart, but for metric spaces there is more data to be given, and it's not irrelevant (different choices produce distinct metric spaces). So I would have this just as a def, and let people instantiate it if they want.

Reply:
I agree. Changed to a def.